File: backend/app/schemas/human_design.py
"""
Human Design Schemas

Pydantic schemas for Human Design calculations and responses.
"""

from pydantic import BaseModel, Field
from typing import Dict, List, Optional, Any
from datetime import datetime
from enum import Enum


# ==============================================================================
# ENUMS
# ==============================================================================

class ZodiacType(str, Enum):
    TROPICAL = "tropical"
    SIDEREAL = "sidereal"


class SiderealMethod(str, Enum):
    SHIFT_POSITIONS = "shift_positions"  # Apply ayanamsa to planet positions
    SHIFT_WHEEL = "shift_wheel"          # Shift gate wheel by ayanamsa


class HDType(str, Enum):
    MANIFESTOR = "Manifestor"
    GENERATOR = "Generator"
    MANIFESTING_GENERATOR = "Manifesting Generator"
    PROJECTOR = "Projector"
    REFLECTOR = "Reflector"


class HDAuthority(str, Enum):
    EMOTIONAL = "Emotional"
    SACRAL = "Sacral"
    SPLENIC = "Splenic"
    EGO_MANIFESTED = "Ego Manifested"
    EGO_PROJECTED = "Ego Projected"
    SELF_PROJECTED = "Self-Projected"
    MENTAL = "Mental/Environment"
    LUNAR = "Lunar"
    NONE = "None"


class HDDefinition(str, Enum):
    NONE = "No Definition"
    SINGLE = "Single Definition"
    SPLIT = "Split Definition"
    TRIPLE_SPLIT = "Triple Split"
    QUADRUPLE_SPLIT = "Quadruple Split"


class CrossType(str, Enum):
    RIGHT_ANGLE = "Right Angle"
    JUXTAPOSITION = "Juxtaposition"
    LEFT_ANGLE = "Left Angle"


# ==============================================================================
# NESTED RESPONSE MODELS
# ==============================================================================

class GateActivation(BaseModel):
    """A single gate activation with full details."""
    gate: int = Field(..., ge=1, le=64, description="Gate number (1-64)")
    line: int = Field(..., ge=1, le=6, description="Line (1-6)")
    color: int = Field(..., ge=1, le=6, description="Color (1-6)")
    tone: int = Field(..., ge=1, le=6, description="Tone (1-6)")
    base: int = Field(..., ge=1, le=5, description="Base (1-5)")

    planet: str = Field(..., description="Planet that activates this gate")
    longitude: float = Field(..., description="Ecliptic longitude in degrees")
    sign: str = Field(..., description="Zodiac sign")
    degree_in_sign: float = Field(..., description="Degree within sign")

    gate_name: str = Field(..., description="Name of the gate")
    gate_keyword: str = Field(..., description="Keyword for the gate")

    class Config:
        json_schema_extra = {
            "example": {
                "gate": 1,
                "line": 4,
                "color": 2,
                "tone": 3,
                "base": 1,
                "planet": "Sun",
                "longitude": 276.5,
                "sign": "Capricorn",
                "degree_in_sign": 6.5,
                "gate_name": "The Creative",
                "gate_keyword": "Expression"
            }
        }


class ChannelDefinition(BaseModel):
    """A defined channel connecting two centers."""
    gate1: int = Field(..., description="First gate of the channel")
    gate2: int = Field(..., description="Second gate of the channel")
    name: str = Field(..., description="Channel name")
    center1: str = Field(..., description="First center")
    center2: str = Field(..., description="Second center")
    circuit: str = Field(..., description="Circuit type")
    description: str = Field(..., description="Channel description")

    # Which activations define this channel
    gate1_activations: List[str] = Field(default_factory=list, description="Planets activating gate1")
    gate2_activations: List[str] = Field(default_factory=list, description="Planets activating gate2")

    # Whether activation is from Personality, Design, or both
    activation_type: str = Field(..., description="'personality', 'design', or 'both'")


class CenterDefinition(BaseModel):
    """Definition status of a center."""
    name: str = Field(..., description="Center name")
    defined: bool = Field(..., description="Whether the center is defined")

    # Which gates are activated in this center
    activated_gates: List[int] = Field(default_factory=list, description="All activated gates")

    # Which channels define this center
    defining_channels: List[str] = Field(default_factory=list, description="Channel names that define this center")

    # Center metadata
    theme: str = Field(..., description="Center's theme")
    biological_correlation: str = Field(..., description="Biological correlation")
    not_self_theme: str = Field(..., description="Not-self theme")


class ProfileInfo(BaseModel):
    """Profile information."""
    personality_line: int = Field(..., ge=1, le=6, description="Personality Sun line")
    design_line: int = Field(..., ge=1, le=6, description="Design Sun line")
    name: str = Field(..., description="Profile name (e.g., '3/5 Martyr/Heretic')")
    angle: str = Field(..., description="Cross angle type")
    description: str = Field(..., description="Profile description")


class IncarnationCross(BaseModel):
    """Incarnation Cross information."""
    name: str = Field(..., description="Cross name")
    cross_type: CrossType = Field(..., description="Right Angle, Juxtaposition, or Left Angle")
    quarter: str = Field(..., description="Quarter of the wheel")

    personality_sun_gate: int = Field(..., description="Personality Sun gate")
    personality_earth_gate: int = Field(..., description="Personality Earth gate")
    design_sun_gate: int = Field(..., description="Design Sun gate")
    design_earth_gate: int = Field(..., description="Design Earth gate")

    description: str = Field(default="", description="Cross description")


class Variables(BaseModel):
    """Advanced: Variables (arrows) for PHS and environment."""
    # Design Sun determines digestion (left variable)
    design_sun_color: int = Field(..., description="Design Sun color")
    design_sun_tone: int = Field(..., description="Design Sun tone")
    digestion: str = Field(..., description="Digestion strategy")

    # Personality Sun determines cognition (right variable)
    personality_sun_color: int = Field(..., description="Personality Sun color")
    personality_sun_tone: int = Field(..., description="Personality Sun tone")
    cognition: str = Field(..., description="Cognitive style")

    # Design Node determines environment
    design_node_color: int = Field(..., description="Design North Node color")
    design_node_tone: int = Field(..., description="Design North Node tone")
    environment: str = Field(..., description="Optimal environment")

    # Personality Node determines perspective
    personality_node_color: int = Field(..., description="Personality North Node color")
    personality_node_tone: int = Field(..., description="Personality North Node tone")
    perspective: str = Field(..., description="Perspective/motivation")

    # Arrow directions (left/right for each)
    left_arrow_1: str = Field(..., description="First left arrow direction")
    left_arrow_2: str = Field(..., description="Second left arrow direction")
    right_arrow_1: str = Field(..., description="First right arrow direction")
    right_arrow_2: str = Field(..., description="Second right arrow direction")


# ==============================================================================
# REQUEST SCHEMAS
# ==============================================================================

class HDCalculationRequest(BaseModel):
    """Request to calculate a Human Design chart."""
    birth_data_id: str = Field(..., description="UUID of the birth data record")

    zodiac_type: ZodiacType = Field(
        default=ZodiacType.TROPICAL,
        description="Zodiac system to use"
    )
    sidereal_method: SiderealMethod = Field(
        default=SiderealMethod.SHIFT_POSITIONS,
        description="Method for sidereal calculation (only used if zodiac_type is sidereal)"
    )
    ayanamsa: str = Field(
        default="lahiri",
        description="Ayanamsa to use for sidereal calculations"
    )

    include_variables: bool = Field(
        default=True,
        description="Whether to include Variables (arrows) calculation"
    )
    chart_name: Optional[str] = Field(
        default=None,
        description="Optional name for the chart"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "birth_data_id": "123e4567-e89b-12d3-a456-426614174000",
                "zodiac_type": "tropical",
                "include_variables": True,
                "chart_name": "My Human Design Chart"
            }
        }


class HDInterpretationRequest(BaseModel):
    """Request for AI interpretation of HD element."""
    chart_id: Optional[str] = Field(None, description="Chart ID for context")
    element_type: str = Field(
        ...,
        description="Type of element: 'type', 'profile', 'channel', 'gate', 'authority', 'full'"
    )
    element_id: Optional[str] = Field(
        None,
        description="Specific element identifier (e.g., gate number, channel name)"
    )
    context: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Additional context for the interpretation"
    )


# ==============================================================================
# RESPONSE SCHEMAS
# ==============================================================================

class HDChartResponse(BaseModel):
    """Complete Human Design chart response."""
    id: Optional[str] = Field(None, description="Chart ID if saved")
    birth_data_id: str = Field(..., description="Birth data ID")

    # Core HD elements
    hd_type: HDType = Field(..., description="Human Design Type")
    strategy: str = Field(..., description="Life strategy")
    authority: HDAuthority = Field(..., description="Inner authority")
    authority_description: str = Field(..., description="Authority guidance")
    signature: str = Field(..., description="Type signature theme")
    not_self: str = Field(..., description="Not-self theme")

    # Profile
    profile: ProfileInfo = Field(..., description="Profile information")

    # Definition
    definition: HDDefinition = Field(..., description="Definition type")
    definition_description: str = Field(..., description="Definition explanation")

    # Planetary activations
    personality_activations: Dict[str, GateActivation] = Field(
        ..., description="Personality (conscious) planetary activations"
    )
    design_activations: Dict[str, GateActivation] = Field(
        ..., description="Design (unconscious) planetary activations"
    )

    # Centers
    centers: Dict[str, CenterDefinition] = Field(..., description="9 centers with definition status")
    defined_centers: List[str] = Field(..., description="List of defined center names")
    undefined_centers: List[str] = Field(..., description="List of undefined center names")

    # Channels
    channels: List[ChannelDefinition] = Field(..., description="Defined channels")

    # All activated gates (for visualization)
    all_activated_gates: List[int] = Field(..., description="All gates with activations")
    personality_gates: List[int] = Field(..., description="Gates activated by Personality")
    design_gates: List[int] = Field(..., description="Gates activated by Design")

    # Incarnation Cross
    incarnation_cross: IncarnationCross = Field(..., description="Incarnation Cross")

    # Variables (optional)
    variables: Optional[Variables] = Field(None, description="Variables/arrows if calculated")

    # Timing information
    personality_datetime: datetime = Field(..., description="Birth datetime (Personality)")
    design_datetime: datetime = Field(..., description="Design datetime (~88 days before)")
    design_days_before: float = Field(..., description="Actual days before birth for Design")

    # Calculation metadata
    calculation_info: Dict[str, Any] = Field(
        default_factory=dict,
        description="Calculation metadata (zodiac, ayanamsa, etc.)"
    )

    created_at: Optional[datetime] = Field(None, description="When chart was created")

    class Config:
        json_schema_extra = {
            "example": {
                "hd_type": "Generator",
                "strategy": "To Respond",
                "authority": "Sacral",
                "profile": {
                    "personality_line": 3,
                    "design_line": 5,
                    "name": "3/5 Martyr/Heretic",
                    "angle": "Right Angle"
                },
                "definition": "Single Definition"
            }
        }


class HDGateInfo(BaseModel):
    """Information about a single gate."""
    number: int
    name: str
    keyword: str
    description: str = ""
    i_ching_name: str
    center: str
    circuit: Optional[str] = None
    channel_partner: Optional[int] = None
    channel_name: Optional[str] = None
    line_descriptions: Dict[int, str] = Field(default_factory=dict)


class HDChannelInfo(BaseModel):
    """Information about a single channel."""
    gate1: int
    gate2: int
    name: str
    center1: str
    center2: str
    circuit: str
    description: str


class HDCenterInfo(BaseModel):
    """Information about a single center."""
    name: str
    biological: str
    theme: str
    function: str
    defined_theme: str
    undefined_theme: str
    not_self: str
    gates: List[int]
    is_motor: bool
    is_pressure: bool


class HDTypeInfo(BaseModel):
    """Information about a Human Design Type."""
    name: str
    strategy: str
    signature: str
    not_self: str
    aura: str
    percentage: str
    description: str


# ==============================================================================
# AI INTERPRETATION RESPONSES
# ==============================================================================

class HDTypeInterpretationResponse(BaseModel):
    """AI interpretation of Type, Strategy, and Authority."""
    hd_type: str
    strategy: str
    authority: str
    interpretation: str


class HDProfileInterpretationResponse(BaseModel):
    """AI interpretation of Profile."""
    profile: str
    interpretation: str


class HDChannelInterpretationResponse(BaseModel):
    """AI interpretation of a Channel."""
    channel_name: str
    gates: List[int]
    interpretation: str


class HDGateInterpretationResponse(BaseModel):
    """AI interpretation of a Gate."""
    gate: int
    gate_name: str
    planet: str
    line: int
    is_personality: bool
    interpretation: str


class HDFullReadingResponse(BaseModel):
    """AI-generated full Human Design reading."""
    chart_id: Optional[str]
    hd_type: str
    profile: str
    authority: str
    definition: str
    reading: str
    sections: Dict[str, str] = Field(
        default_factory=dict,
        description="Reading broken into sections (type, profile, centers, channels, guidance)"
    )
    generated_at: datetime = Field(default_factory=datetime.utcnow)


# ==============================================================================
# LIST RESPONSES
# ==============================================================================

class HDGatesListResponse(BaseModel):
    """Response for listing all gates."""
    gates: List[HDGateInfo]
    count: int = 64


class HDChannelsListResponse(BaseModel):
    """Response for listing all channels."""
    channels: List[HDChannelInfo]
    count: int = 36


class HDCentersListResponse(BaseModel):
    """Response for listing all centers."""
    centers: List[HDCenterInfo]
    count: int = 9


class HDTypesListResponse(BaseModel):
    """Response for listing all types."""
    types: List[HDTypeInfo]
    count: int = 5

File: frontend/src/lib/api/humanDesign.ts
/**
 * Human Design API client
 *
 * Part of Phase 4: Human Design System Integration
 */
import { apiClient, getErrorMessage } from './client'

// ==================== Type Definitions ====================

export interface GateActivation {
  gate: number
  line: number
  color: number
  tone: number
  base: number
  planet: string
  longitude: number
  sign: string
  degree_in_sign: number
  degree?: number  // Alias for degree_in_sign
  gate_name: string
  gate_keyword: string
  is_design?: boolean
}

export interface ChannelDefinition {
  channel_id?: string
  name: string
  gate1: number
  gate2: number
  gates?: [number, number]  // Computed from gate1, gate2
  center1: string
  center2: string
  center_from?: string  // Alias for center1
  center_to?: string    // Alias for center2
  circuit: string
  type?: string
  description: string
  keynote?: string
  gate1_activations: string[]
  gate2_activations: string[]
  activation_type: string
}

export interface CenterDefinition {
  center?: string  // Added by transform from dict key
  name: string
  defined: boolean
  activated_gates: number[]
  gates?: number[]  // Alias for activated_gates
  active_gates?: number[]  // Alias for activated_gates
  defining_channels: string[]
  biological_correlation: string
  theme: string
  not_self_theme: string
}

export interface ProfileInfo {
  profile: string
  name: string
  personality_line: number
  design_line: number
  angle: string
  description: string
}

export interface IncarnationCross {
  name: string
  type: string
  gates: {
    personality_sun: number
    personality_earth: number
    design_sun: number
    design_earth: number
  }
  description: string
}

export interface Variables {
  determination: {
    color: number
    tone: number
    base: number
    arrow: string
    description: string
  }
  environment: {
    color: number
    tone: number
    base: number
    arrow: string
    description: string
  }
  perspective: {
    color: number
    tone: number
    base: number
    arrow: string
    description: string
  }
  awareness: {
    color: number
    tone: number
    base: number
    arrow: string
    description: string
  }
}

export interface HDChartResponse {
  id: string | null
  birth_data_id: string

  // Core HD attributes - backend uses hd_type
  hd_type: string
  type?: string  // Alias for hd_type, populated by transform
  strategy: string
  authority: string
  authority_description: string
  signature: string
  not_self: string
  definition: string
  definition_description: string
  profile: ProfileInfo
  incarnation_cross: IncarnationCross

  // Activations - backend returns as dict keyed by planet
  personality_activations: Record<string, GateActivation> | GateActivation[]
  design_activations: Record<string, GateActivation> | GateActivation[]
  all_activated_gates: number[]
  personality_gates: number[]
  design_gates: number[]

  // Structure - backend returns centers as dict, channels as array
  channels: ChannelDefinition[]
  centers: Record<string, CenterDefinition> | CenterDefinition[]
  defined_centers: string[]
  undefined_centers: string[]

  // Advanced
  variables?: Variables

  // Timestamps
  personality_datetime: string
  design_datetime: string
  design_days_before: number

  // Calculation info
  calculation_info: {
    zodiac_type: string
    sidereal_method?: string
    ayanamsa?: string
    ayanamsa_value?: number
  }
  created_at: string
}

// Reference data types
export interface HDGateInfo {
  gate: number
  name: string
  center: string
  keynote: string
  description: string
  theme: string
  channel_partners: number[]
  hexagram_name: string
  shadow: string
  gift: string
  siddhi: string
  // Additional fields from backend
  channel_name?: string
  line_descriptions?: Record<string, string>
}

// Raw response from backend uses gate1/gate2, center1/center2
interface HDChannelInfoRaw {
  gate1: number
  gate2: number
  name: string
  center1: string
  center2: string
  circuit: string
  description: string
}

// Transformed for frontend use
export interface HDChannelInfo {
  channel_id: string
  name: string
  gates: [number, number]
  center_from: string
  center_to: string
  type: string
  keynote: string
  description: string
}

export interface HDCenterInfo {
  center: string
  biological_correlation: string
  theme: string
  gates: number[]
  defined_meaning: string
  undefined_meaning: string
  not_self_question: string
}

export interface HDTypeInfo {
  type: string
  strategy: string
  signature: string
  not_self_theme: string
  aura: string
  percentage: string
  description: string
}

// API Response types
export interface HDGatesListResponse {
  gates: HDGateInfo[]
  count: number
}

export interface HDChannelsListResponse {
  channels: HDChannelInfo[]
  count: number
}

export interface HDCentersListResponse {
  centers: HDCenterInfo[]
  count: number
}

export interface HDTypesListResponse {
  types: HDTypeInfo[]
  count: number
}

// Interpretation response types
export interface HDTypeInterpretationResponse {
  type: string
  interpretation: string
}

export interface HDProfileInterpretationResponse {
  profile: string
  interpretation: string
}

export interface HDChannelInterpretationResponse {
  channel_id: string
  interpretation: string
}

export interface HDGateInterpretationResponse {
  gate: number
  interpretation: string
}

export interface HDFullReadingResponse {
  birth_data_id: string
  reading: string
  sections: {
    type_strategy: string
    authority: string
    profile: string
    definition: string
    channels: string
    incarnation_cross: string
    centers: string
    variables?: string
  }
}

// ==================== API Functions ====================

/**
 * Transform backend response to normalized frontend format
 */
function transformChartResponse(data: HDChartResponse): HDChartResponse {
  // Add type alias for hd_type
  const transformed: HDChartResponse = {
    ...data,
    type: data.hd_type,
  }

  // Convert personality_activations dict to array if needed
  if (data.personality_activations && !Array.isArray(data.personality_activations)) {
    transformed.personality_activations = Object.values(data.personality_activations).map(act => ({
      ...act,
      degree: act.degree_in_sign,
    }))
  }

  // Convert design_activations dict to array if needed
  if (data.design_activations && !Array.isArray(data.design_activations)) {
    transformed.design_activations = Object.values(data.design_activations).map(act => ({
      ...act,
      degree: act.degree_in_sign,
    }))
  }

  // Convert centers dict to array if needed with aliases
  if (data.centers && !Array.isArray(data.centers)) {
    transformed.centers = Object.entries(data.centers).map(([key, value]) => ({
      ...value,
      center: key,
      gates: value.activated_gates,
      active_gates: value.activated_gates,
    }))
  }

  // Add gates tuple to channels
  if (Array.isArray(data.channels)) {
    transformed.channels = data.channels.map(ch => ({
      ...ch,
      gates: [ch.gate1, ch.gate2] as [number, number],
      channel_id: `${ch.gate1}-${ch.gate2}`,
      center_from: ch.center1,
      center_to: ch.center2,
    }))
  }

  return transformed
}

/**
 * Calculate Human Design chart for birth data
 */
export async function calculateHDChart(
  birthDataId: string,
  options: {
    zodiac?: 'tropical' | 'sidereal'
    sidereal_method?: 'shift_positions' | 'shift_wheel'
    ayanamsa?: string
    include_variables?: boolean
  } = {}
): Promise<HDChartResponse> {
  try {
    const params = new URLSearchParams()
    if (options.zodiac) params.append('zodiac', options.zodiac)
    if (options.sidereal_method) params.append('sidereal_method', options.sidereal_method)
    if (options.ayanamsa) params.append('ayanamsa', options.ayanamsa)
    if (options.include_variables !== undefined) {
      params.append('include_variables', String(options.include_variables))
    }

    const url = `/human-design/calculate/${birthDataId}${params.toString() ? '?' + params.toString() : ''}`
    // Use longer timeout for HD calculations (30 seconds)
    const response = await apiClient.get(url, { timeout: 30000 })
    return transformChartResponse(response.data)
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Calculate HD chart via POST with full options
 */
export async function calculateHDChartPost(
  birthDataId: string,
  options: {
    zodiac?: 'tropical' | 'sidereal'
    sidereal_method?: 'shift_positions' | 'shift_wheel'
    ayanamsa?: string
    include_variables?: boolean
    custom_design_datetime?: string
  } = {}
): Promise<HDChartResponse> {
  try {
    // Use longer timeout for HD calculations (30 seconds)
    const response = await apiClient.post('/human-design/calculate', {
      birth_data_id: birthDataId,
      ...options
    }, { timeout: 30000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== Reference Data ====================

/**
 * Get list of all 64 gates
 */
export async function getGatesList(): Promise<HDGatesListResponse> {
  try {
    const response = await apiClient.get('/human-design/gates')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// Raw response from backend for gates
interface HDGateInfoRaw {
  number: number
  name: string
  keyword: string
  description: string
  i_ching_name: string
  center: string
  circuit: string | null
  channel_partner: number
  channel_name: string
  line_descriptions: Record<string, string>
}

/**
 * Get info for a specific gate
 * Transforms backend response to frontend format
 */
export async function getGateInfo(gateNumber: number): Promise<HDGateInfo> {
  try {
    const response = await apiClient.get<HDGateInfoRaw>(`/human-design/gates/${gateNumber}`)
    const raw = response.data
    // Transform backend response to frontend format
    return {
      gate: raw.number,
      name: raw.name,
      center: raw.center,
      keynote: raw.keyword,
      description: raw.description || '',
      theme: raw.circuit || '',
      channel_partners: raw.channel_partner ? [raw.channel_partner] : [],
      hexagram_name: raw.i_ching_name,
      // Gene Keys not available from backend yet
      shadow: '',
      gift: '',
      siddhi: '',
      // Additional fields for display
      channel_name: raw.channel_name,
      line_descriptions: raw.line_descriptions,
    }
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get list of all 36 channels
 */
export async function getChannelsList(): Promise<HDChannelsListResponse> {
  try {
    const response = await apiClient.get('/human-design/channels')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get info for a specific channel
 * Transforms backend response (gate1/gate2, center1/center2) to frontend format
 */
export async function getChannelInfo(channelId: string): Promise<HDChannelInfo> {
  try {
    const response = await apiClient.get<HDChannelInfoRaw>(`/human-design/channels/${channelId}`)
    const raw = response.data
    // Transform backend response to frontend format
    return {
      channel_id: `${raw.gate1}-${raw.gate2}`,
      name: raw.name,
      gates: [raw.gate1, raw.gate2],
      center_from: raw.center1,
      center_to: raw.center2,
      type: raw.circuit,
      keynote: raw.description,
      description: raw.description,
    }
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get list of all 9 centers
 */
export async function getCentersList(): Promise<HDCentersListResponse> {
  try {
    const response = await apiClient.get('/human-design/centers')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get info for a specific center
 */
export async function getCenterInfo(centerName: string): Promise<HDCenterInfo> {
  try {
    const response = await apiClient.get(`/human-design/centers/${centerName}`)
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get list of all HD types
 */
export async function getTypesList(): Promise<HDTypesListResponse> {
  try {
    const response = await apiClient.get('/human-design/types')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get info for a specific type
 */
export async function getTypeInfo(typeName: string): Promise<HDTypeInfo> {
  try {
    const response = await apiClient.get(`/human-design/types/${typeName}`)
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== AI Interpretations ====================

/**
 * Get AI interpretation for HD type
 */
export async function getTypeInterpretation(
  typeName: string,
  strategy: string,
  signature: string
): Promise<HDTypeInterpretationResponse> {
  try {
    const response = await apiClient.post('/human-design/ai/interpret-type', {
      type: typeName,
      strategy,
      signature
    })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get AI interpretation for HD profile
 */
export async function getProfileInterpretation(
  profile: string,
  angle: string
): Promise<HDProfileInterpretationResponse> {
  try {
    const response = await apiClient.post('/human-design/ai/interpret-profile', {
      profile,
      angle
    })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get AI interpretation for HD channel
 */
export async function getChannelInterpretation(
  channelId: string,
  channelName: string,
  keynote: string
): Promise<HDChannelInterpretationResponse> {
  try {
    const response = await apiClient.post('/human-design/ai/interpret-channel', {
      channel_id: channelId,
      name: channelName,
      keynote
    })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get AI interpretation for HD gate
 */
export async function getGateInterpretation(
  gate: number,
  line: number,
  planet: string,
  isDesign: boolean
): Promise<HDGateInterpretationResponse> {
  try {
    const response = await apiClient.post('/human-design/ai/interpret-gate', {
      gate,
      line,
      planet,
      is_design: isDesign
    })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get full AI reading for HD chart
 */
export async function getFullHDReading(
  birthDataId: string,
  options: {
    zodiac?: 'tropical' | 'sidereal'
    sidereal_method?: 'shift_positions' | 'shift_wheel'
    ayanamsa?: string
    include_variables?: boolean
  } = {}
): Promise<HDFullReadingResponse> {
  try {
    // Use longer timeout for AI reading generation (60 seconds)
    const response = await apiClient.post('/human-design/ai/full-reading', {
      birth_data_id: birthDataId,
      ...options
    }, { timeout: 60000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== Helper Functions ====================

/**
 * Get display name for HD type
 */
export function getTypeDisplayName(type: string): string {
  const names: Record<string, string> = {
    'generator': 'Generator',
    'manifesting_generator': 'Manifesting Generator',
    'projector': 'Projector',
    'manifestor': 'Manifestor',
    'reflector': 'Reflector'
  }
  return names[type.toLowerCase()] || type
}

/**
 * Get color for HD type
 */
export function getTypeColor(type: string): string {
  const colors: Record<string, string> = {
    'generator': '#E6B800',        // Gold/Yellow
    'manifesting_generator': '#FF6B35', // Orange
    'projector': '#3B82F6',        // Blue
    'manifestor': '#EF4444',       // Red
    'reflector': '#A855F7'         // Purple
  }
  return colors[type.toLowerCase()] || '#6B7280'
}

/**
 * Get center color based on defined state
 */
export function getCenterColor(defined: boolean): string {
  return defined ? '#E6B800' : '#FFFFFF' // Gold if defined, white if undefined
}

/**
 * Get center display name
 */
export function getCenterDisplayName(center: string): string {
  const names: Record<string, string> = {
    'head': 'Head',
    'ajna': 'Ajna',
    'throat': 'Throat',
    'g_center': 'G Center',
    'heart': 'Heart/Ego',
    'sacral': 'Sacral',
    'solar_plexus': 'Solar Plexus',
    'spleen': 'Spleen',
    'root': 'Root'
  }
  return names[center.toLowerCase()] || center
}

/**
 * Get authority display name
 */
export function getAuthorityDisplayName(authority: string): string {
  const names: Record<string, string> = {
    'emotional': 'Emotional/Solar Plexus',
    'sacral': 'Sacral',
    'splenic': 'Splenic',
    'ego_projected': 'Ego Projected',
    'ego_manifested': 'Ego Manifested',
    'self_projected': 'Self-Projected',
    'mental': 'Mental/Environmental',
    'lunar': 'Lunar',
    'none': 'No Authority'
  }
  return names[authority.toLowerCase()] || authority
}

/**
 * Get definition type display name
 */
export function getDefinitionDisplayName(definition: string): string {
  const names: Record<string, string> = {
    'single': 'Single Definition',
    'split': 'Split Definition',
    'triple_split': 'Triple Split Definition',
    'quadruple_split': 'Quadruple Split Definition',
    'no_definition': 'No Definition'
  }
  return names[definition.toLowerCase()] || definition
}

/**
 * Format gate with line (e.g., "Gate 1.3")
 */
export function formatGateLine(gate: number, line: number): string {
  return `Gate ${gate}.${line}`
}

/**
 * Get planet symbol for HD activations
 */
export function getPlanetSymbol(planet: string): string {
  const symbols: Record<string, string> = {
    'Sun': '☉',
    'Earth': '⊕',
    'Moon': '☽',
    'North Node': '☊',
    'South Node': '☋',
    'Mercury': '☿',
    'Venus': '♀',
    'Mars': '♂',
    'Jupiter': '♃',
    'Saturn': '♄',
    'Uranus': '♅',
    'Neptune': '♆',
    'Pluto': '♇'
  }
  return symbols[planet] || planet
}

/**
 * Get variable arrow meaning
 */
export function getVariableArrowMeaning(position: string, arrow: string): string {
  const meanings: Record<string, Record<string, string>> = {
    'determination': {
      'left': 'Active/Strategic eating',
      'right': 'Passive/Receptive eating'
    },
    'environment': {
      'left': 'Active environment selection',
      'right': 'Passive environment recognition'
    },
    'perspective': {
      'left': 'Focused/Active perception',
      'right': 'Peripheral/Passive perception'
    },
    'awareness': {
      'left': 'Strategic awareness',
      'right': 'Receptive awareness'
    }
  }
  return meanings[position]?.[arrow] || `${arrow} arrow`
}

/**
 * Check if gate pair forms a channel
 */
export function formsChannel(gate1: number, gate2: number, channels: ChannelDefinition[]): boolean {
  return channels.some(ch => {
    const gates = ch.gates || (ch.gate1 !== undefined && ch.gate2 !== undefined ? [ch.gate1, ch.gate2] : null)
    if (!gates) return false
    return (gates[0] === gate1 && gates[1] === gate2) ||
           (gates[0] === gate2 && gates[1] === gate1)
  })
}

/**
 * Get all gates for a center
 */
export function getGatesForCenter(centerName: string, centers: CenterDefinition[]): number[] {
  const center = centers.find(c => {
    const name = c.center || c.name || ''
    return name.toLowerCase() === centerName.toLowerCase()
  })
  return center?.gates || center?.active_gates || center?.activated_gates || []
}

/**
 * Calculate percentage of centers defined
 */
export function getDefinitionPercentage(centers: CenterDefinition[]): number {
  const defined = centers.filter(c => c.defined).length
  return Math.round((defined / centers.length) * 100)
}
